rclrs/src/future.rs

```diff
@@ -0,0 +1,63 @@
+/// Based on https://www.viget.com/articles/understanding-futures-in-rust-part-1/
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Context;
+use std::task::Poll;
+use std::task::RawWaker;
+use std::task::RawWakerVTable;
+use std::task::Waker;
+
+#[derive(Default)]
+pub struct RclFuture<T> {
+    value: Option<T>,
+}
+
+impl<T: Default + Clone> RclFuture<T> {
+    pub fn new() -> RclFuture<T> {
+        Self { value: None }
+    }
+
+    pub fn set_value(&mut self, msg: T) {
+        self.value = Some(msg);
+    }
+}
+
+impl<T: Clone> Future for RclFuture<T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, _ctx: &mut Context) -> Poll<Self::Output> {
+        if let Some(value) = &self.value {
+            Poll::Ready(value.clone())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RclWaker {}
```

Comment:
I don't quite understand this part, at least the waker seems to have no functionality.

`r2r` and `rclrust` both implement the client future by storing one side of a `oneshot::channel` (that is itself a `Future`) inside the client, and the other inside the  response future. Would something like that maybe be more straightforward?

Reply:
Yeah, for now the waker is just boilerplate so that `RclFuture` can implement the `Future` trait. The rationale for this is such that once we support ROS2 executors, the `RclFuture` can take reference to the current executor and use it to resolve any pending computation. But your suggestion of using channels is interesting, thanks. I'll explore that, it might be clearer and easier to read than implementing a custom `Future`

Reply:
Well, a waker is not required to implement `Future`, it's required for an async executor like `tokio` (and the `spin_until_future_complete` function is kind of an async executor, but only for `RclFuture`).

But my proposal would be, let's first get an MVP of clients merged that doesn't use futures at all, but only callbacks.
After that, let's do a follow-up for futures. The reason being that I don't have the context behind `rclcpp::spin_until_future_complete`, and futures in C++ in general yet to understand what led to this design. Rust's async story seems very different from C++'s, so I can see this being a wider discussion than appropriate for an (already large) PR.

Reply:
> Well, a waker is not required to implement Future, it's required for an async executor like tokio (and the spin_until_future_complete function is kind of an async executor, but only for RclFuture).

It is required if you want `.wait` to actually do anything, but that's semantics.

> But my proposal would be, let's first get an MVP of clients merged that doesn't use futures at all, but only callbacks.

That's why I added `async_send_request_with_callback`, which just accepts a callback. I'm not too happy with the future stuff in this PR, but I wanted to get something in that felt familiar to `rclcpp` users.

> The reason being that I don't have the context behind `rclcpp::spin_until_future_complete`

The rationale for `rclcpp::spin_until_future_complete` is that the only way for getting the value out of a future is to call `.get`, which blocks and unless you add threads to the mix, it won't do anything. Given that the first ROS2 executor was single threaded (and it is in fact, the default one), this model wouldn't work, so what `rclcpp::spin_until_future_complete` does is it waits for it to be resolved for a certain time, if it's not resolved, call spin, wait again, etc. until it's resolved and so calling `.get` won't block and will return the value to the caller.

Reply:
Ok. I'd guess `rclcpp` users will probably be fine with just a callback API at first – @i-am-chauhan, as a requester of this feature, let us know if you have any opinion on this. In any case, I'm not suggesting to keep the "future" stuff out of the crates.io release entirely, only out of this PR.

If you're against that suggestion, as a compromise, we could also keep `RclFuture` in this PR, but not implement `Future` on it. The trait is not actually necessary (so you can just implement `poll()` on the type, and it doesn't even need `task::Context`), and might lead people to believe they can use it with a `tokio` runtime, which is not the case.

Reply:
I've reworked the futures stuff to use `Future`s from `future-rs` and tokio as the runtime in the examples. I haven't looked into `r2r` yet, but from what I've seen, `rclrust` has implemented executors incorrectly, as each node has its own executor thread which spins (https://github.com/rclrust/rclrust/blob/main/rclrust/src/node.rs#L219-L222=), with this model, you'd have as many threads as nodes.

The `RclFuture`/`rclrs::spin_until_future_complete` implementation has the advantage that no threads are used anywhere. With the latest changes, there's only one thread, that simply calls `rclrs::spin`. With ROS2's executor API, we could hide the thread logic and make it as seamless as possible, but for now we only have `rclrs::spin*`

I still need to do some cleanups, but with the latest changes the API may be easier to understand for people who are used to the Rust async API, but not the ROS2 API.

Reply:
It doesn't matter whether `Future` is from `futures` or `std::task`, `futures` just re-exports the latter one.

Your `Future` implementation is also incorrect, I think. As mentioned above, the waker is provided by the async executor, not the implementation of a future, so no need to define one at all. And the future should, upon first being polled, store the waker so as to notify the async executor when it's ready (i.e. the future would need to give the waker to a wait set, and the wait set needs to use it to wake up the future when the response arrives).

I've started a discussion in the Matrix chat about this, since I think it will be more efficient..

Reply:
If look at the latest commits, RclFuture is not use anywhere any more, as I said, there were still some cleanups pending, one of them is removing RclFuture from this PR.

> Your Future implementation is also incorrect, I think. As mentioned above, the waker is provided by the async executor, not the implementation of a future, so no need to define one at all.

And therefore, that's why there's a waker for the ROS2 executor model. Anyway, is it worth discussing this further? I've already replaced RclFuture with Rust's futures.

> I've started a discussion in the Matrix chat about this, since I think it will be more efficient..

Please don't do that in the future, I prefer not moving this discussion to the Matrix channel, it's much less efficient. If you want to have a more general futures vs callbacks discussion, please create a discussion on GitHub so that there's a trail of whatever decisions we make.

Reply:
I wasn't aware you were removing `RclFuture`. But do you intend to keep `spin_until_future_complete`?

Reply:
I don't see the point of keeping `spin_until_future_complete` if `RclFuture` is not included, the former only works with `RclFuture`s